cpp/modmesh/buffer/BufferBase.hpp

```diff
@@ -0,0 +1,110 @@
+ *   without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <modmesh/base.hpp>
+
+namespace modmesh
+{
+
+/// Base class for buffer-like objects.
+template <typename Derived>
+class BufferBase : public std::enable_shared_from_this<BufferBase<Derived>>
+{
+public:
+
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+
+    explicit operator bool() const { return bool(m_begin); }
+    size_type size() const
+    {
+        return static_cast<size_type>(this->m_end - this->m_begin);
+    }
+    size_t nbytes() const { return size() * sizeof(int8_t); }
+
+    int8_t operator[](size_t it) const { return data(it); }
+    int8_t & operator[](size_t it) { return data(it); }
+
+    int8_t at(size_t it) const
+    {
+        validate_range(it);
+        return data(it);
+    }
+
+    int8_t & at(size_t it)
+    {
+        validate_range(it);
+        return data(it);
+    }
+
+    using iterator = int8_t *;
+    using const_iterator = int8_t const *;
+
+    iterator begin() noexcept { return m_begin; }
+    iterator end() noexcept { return m_end; }
+    const_iterator begin() const noexcept { return m_begin; }
+    const_iterator end() const noexcept { return m_end; }
+    const_iterator cbegin() const noexcept { return m_begin; }
+    const_iterator cend() const noexcept { return m_end; }
+
+    /* Backdoor */
+    int8_t data(size_type it) const { return data()[it]; }
+    int8_t & data(size_type it) { return data()[it]; }
+    int8_t const * data() const noexcept { return data<int8_t>(); }
+    int8_t * data() noexcept { return data<int8_t>(); }
+
+    // clang-format off
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+    template <typename T> T const * data() const noexcept { return reinterpret_cast<T *>(m_begin); }
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+    template <typename T> T * data() noexcept { return reinterpret_cast<T *>(m_begin); }
+    // clang-format on
+
+    constexpr const char * name() const
+    {
+        return static_cast<const Derived *>(this)->name();
+    }
+
+protected:
+    // virtual ~BufferBase() = default;
```

Comment:
@yungyuc  if uncommenting this line, the code can build.

if without the polymorphism, the build encounter the error related to `dynamic_cast`, which seems weird to me, could you take a look and provide some hints?

```log
[ 88%] Building CXX object cpp/modmesh/CMakeFiles/modmesh_primary.dir/testhelper/pymod/wrap_TestSimpleArrayHelper.cpp.o
In file included from /usr/include/c++/10/memory:84,
                 from /usr/local/include/pybind11/detail/common.h:315,
                 from /usr/local/include/pybind11/attr.h:13,
                 from /usr/local/include/pybind11/detail/class.h:12,
                 from /usr/local/include/pybind11/pybind11.h:13,
                 from /home/mujin/modmesh/cpp/modmesh/buffer/pymod/buffer_pymod.hpp:31,
                 from /home/mujin/modmesh/cpp/modmesh/buffer/pymod/wrap_ConcreteBuffer.cpp:29:
/usr/include/c++/10/bits/shared_ptr.h: In instantiation of ‘std::shared_ptr<_Tp> std::dynamic_pointer_cast(const std::shared_ptr<_Tp>&) [with _Tp = modmesh::ConcreteBuffer; _Up = modmesh::BufferBase<modmesh::ConcreteBuffer>]’:
/usr/local/include/pybind11/pybind11.h:1858:80:   required from ‘static void pybind11::class_<type_, options>::init_holder(pybind11::detail::instance*, pybind11::detail::value_and_holder&, const holder_type*, const std::enable_shared_from_this<T>*) [with T = modmesh::BufferBase<modmesh::ConcreteBuffer>; type_ = modmesh::ConcreteBuffer; options = {std::shared_ptr<modmesh::ConcreteBuffer>}; pybind11::class_<type_, options>::holder_type = std::shared_ptr<modmesh::ConcreteBuffer>]’
/usr/local/include/pybind11/pybind11.h:1910:20:   required from ‘static void pybind11::class_<type_, options>::init_instance(pybind11::detail::instance*, const void*) [with type_ = modmesh::ConcreteBuffer; options = {std::shared_ptr<modmesh::ConcreteBuffer>}]’
/usr/local/include/pybind11/pybind11.h:1584:30:   required from ‘pybind11::class_<type_, options>::class_(pybind11::handle, const char*, const Extra& ...) [with Extra = {const char*, pybind11::buffer_protocol}; type_ = modmesh::ConcreteBuffer; options = {std::shared_ptr<modmesh::ConcreteBuffer>}]’
/home/mujin/modmesh/cpp/modmesh/python/common.hpp:394:45:   required from ‘modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::WrapBase(pybind11::module&, const char*, const char*, const Extra& ...) [with Extra = {pybind11::buffer_protocol}; Wrapper = modmesh::python::WrapConcreteBuffer; Wrapped = modmesh::ConcreteBuffer; Holder = std::shared_ptr<modmesh::ConcreteBuffer>; WrappedBase = modmesh::ConcreteBuffer; pybind11::module = pybind11::module_]’
/home/mujin/modmesh/cpp/modmesh/buffer/pymod/wrap_ConcreteBuffer.cpp:49:69:   required from here
/usr/include/c++/10/bits/shared_ptr.h:602:23: error: cannot ‘dynamic_cast’ ‘(& __r)->std::shared_ptr<modmesh::BufferBase<modmesh::ConcreteBuffer> >::<anonymous>.std::__shared_ptr<modmesh::BufferBase<modmesh::ConcreteBuffer>, __gnu_cxx::_S_atomic>::get()’ (of type ‘using element_type = std::remove_extent<modmesh::BufferBase<modmesh::ConcreteBuffer> >::type*’ {aka ‘class modmesh::BufferBase<modmesh::ConcreteBuffer>*’}) to type ‘using element_type = using element_type = std::remove_extent<modmesh::ConcreteBuffer>::type*’ {aka ‘class modmesh::ConcreteBuffer*’} (source type is not polymorphic)
  602 |       if (auto* __p = dynamic_cast<typename _Sp::element_type*>(__r.get()))
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/10/bits/shared_ptr.h: In instantiation of ‘std::shared_ptr<_Tp> std::dynamic_pointer_cast(const std::shared_ptr<_Tp>&) [with _Tp = modmesh::BufferExpander; _Up = modmesh::BufferBase<modmesh::BufferExpander>]’:
/usr/local/include/pybind11/pybind11.h:1858:80:   required from ‘static void pybind11::class_<type_, options>::init_holder(pybind11::detail::instance*, pybind11::detail::value_and_holder&, const holder_type*, const std::enable_shared_from_this<T>*) [with T = modmesh::BufferBase<modmesh::BufferExpander>; type_ = modmesh::BufferExpander; options = {std::shared_ptr<modmesh::BufferExpander>}; pybind11::class_<type_, options>::holder_type = std::shared_ptr<modmesh::BufferExpander>]’
/usr/local/include/pybind11/pybind11.h:1910:20:   required from ‘static void pybind11::class_<type_, options>::init_instance(pybind11::detail::instance*, const void*) [with type_ = modmesh::BufferExpander; options = {std::shared_ptr<modmesh::BufferExpander>}]’
/usr/local/include/pybind11/pybind11.h:1584:30:   required from ‘pybind11::class_<type_, options>::class_(pybind11::handle, const char*, const Extra& ...) [with Extra = {const char*}; type_ = modmesh::BufferExpander; options = {std::shared_ptr<modmesh::BufferExpander>}]’
/home/mujin/modmesh/cpp/modmesh/python/common.hpp:394:45:   required from ‘modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::WrapBase(pybind11::module&, const char*, const char*, const Extra& ...) [with Extra = {}; Wrapper = modmesh::python::WrapBufferExpander; Wrapped = modmesh::BufferExpander; Holder = std::shared_ptr<modmesh::BufferExpander>; WrappedBase = modmesh::BufferExpander; pybind11::module = pybind11::module_]’
/home/mujin/modmesh/cpp/modmesh/buffer/pymod/wrap_ConcreteBuffer.cpp:124:40:   required from here
/usr/include/c++/10/bits/shared_ptr.h:602:23: error: cannot ‘dynamic_cast’ ‘(& __r)->std::shared_ptr<modmesh::BufferBase<modmesh::BufferExpander> >::<anonymous>.std::__shared_ptr<modmesh::BufferBase<modmesh::BufferExpander>, __gnu_cxx::_S_atomic>::get()’ (of type ‘using element_type = std::remove_extent<modmesh::BufferBase<modmesh::BufferExpander> >::type*’ {aka ‘class modmesh::BufferBase<modmesh::BufferExpander>*’}) to type ‘using element_type = using element_type = std::remove_extent<modmesh::BufferExpander>::type*’ {aka ‘class modmesh::BufferExpander*’} (source type is not polymorphic)
gmake[4]: *** [cpp/modmesh/CMakeFiles/modmesh_primary.dir/build.make:118: cpp/modmesh/CMakeFiles/modmesh_primary.dir/buffer/pymod/wrap_ConcreteBuffer.cpp.o] Error 1
gmake[4]: *** Waiting for unfinished jobs....
```

Reply:
the build result can be checked in https://github.com/solvcon/modmesh/pull/363/commits/45ce936eb8ea2963420deee3937ec292180773f7 and https://github.com/solvcon/modmesh/pull/363/commits/e5b8d34a321560f9052c79c6fd5ba772bfeb4467

Reply:
figured out, we cannot put `enable_shared_from_this` in the base class.

Reply:
If we want to put enable_shared_from_this  in the base class, then polymorphism is needed.